common/src/graph_ray_casting.rs

```diff
@@ -0,0 +1,88 @@
+use crate::{
+    chunk_ray_casting::chunk_ray_cast,
+    collision_math::Ray,
+    graph::Graph,
+    node::{Chunk, ChunkId, Coords},
+    proto::Position,
+    traversal::RayTraverser,
+};
+
+/// Performs ray casting against the voxels in the `DualGraph`
+///
+/// The `ray` parameter and any resulting hit normals are given in the local coordinate system of `position`.
+///
+/// The `tanh_distance` is the hyperbolic tangent of the cast_distance, or the distance along the ray to check for hits.
+///
+/// This function may return a `RayCastError` if not enough chunks are generated, even if the ray never reaches an
+/// ungenerated chunk. To prevent these errors, make sure that the distance between the ray's start point and the center of
+/// the closest node with ungenerated chunks is greater than `cast_distance + dodeca::BOUNDING_SPHERE_RADIUS`
+pub fn ray_cast(
+    graph: &Graph,
+    position: &Position,
+    ray: &Ray,
+    mut tanh_distance: f32,
+) -> Result<Option<GraphCastHit>, RayCastError> {
+    // A ray cast is assumed to be a miss until a collision is found.
+    // This `hit` variable gets updated over time before being returned.
+    let mut hit: Option<GraphCastHit> = None;
+
+    let mut traverser = RayTraverser::new(graph, *position, ray, 0.0);
+    while let Some((chunk, transform)) = traverser.next(tanh_distance) {
+        let Some(chunk) = chunk else {
+            // Ray reached chunk outside of graph
+            return Err(RayCastError::OutOfBounds);
+        };
+        let Chunk::Populated {
+            voxels: ref voxel_data,
+            ..
+        } = graph[chunk]
+        else {
+            // Ray reached unpopulated chunk
+            return Err(RayCastError::OutOfBounds);
+        };
+
+        hit = chunk_ray_cast(
+            voxel_data,
+            graph.layout(),
+            &(transform * ray),
+            tanh_distance,
+        )
+        .map_or(hit, |hit| {
+            tanh_distance = hit.tanh_distance;
+            Some(GraphCastHit {
+                tanh_distance: hit.tanh_distance,
+                chunk,
+                voxel_coords: hit.voxel_coords,
+                face_axis: hit.face_axis,
+                face_direction: hit.face_direction,
+            })
+        });
+    }
+
+    Ok(hit)
+}
+
+#[derive(Debug)]
+pub enum RayCastError {
+    OutOfBounds,
+}
+
+/// Information about the intersection at the end of a ray segment.
+#[derive(Debug)]
+pub struct GraphCastHit {
```

Comment:
I'm not sure. I think it would be nice for outside code to be able to act as though `chunk_ray_casting.rs` doesn't exist at all, as if it's a private module, instead relying on `graph_ray_casting.rs` exclusively.

In `graph_collision.rs` and `chunk_collision.rs`, the `normal` vector in their respective `GraphCastHit` and `ChunkCastHit` are in different coordinate systems, so the same refactor wouldn't be able to be done there.

There seems to be a bit of a tension between DRY and encapsulation, so I will think on this a bit further.

Reply:
In that case, I'll leave this code as-is for now. It seems worth revisiting if another shape has to be cast in the future.